process/HOPE-3--Dependency-Injection.md

```diff
@@ -0,0 +1,169 @@
+- Directives aren't nestable.
+
+## Proposed Dependency Injection System
+
+This HOPE proposes a new dependency injection system with an aim for clarity and reuse.
+The API of the new system would consist of a `provides` decorator for defining available dependencies, and a `requires` function for using them:
+
+```python
+@provides(name: str = None, api: hug.API = None, override: bool = False)
+
+def requires(name_or_function: Union[str, Callable], *args, **kwargs) -> Any
+```
+
+These two calls are meant to replace the old `Directive` system:
+
+```python
+@directive(apply_globally: bool = False, api: hug.API = None)
+```
+
+Which was used in the following manner:
+
+```python
+@hug.get()
+def my_endpoint(hug_timer):
+    pass
+
+
+# OR in the more modern form:
+
+
+@hug.get()
+def my_endpoint(timer: hug.directives.timer):
+    pass
+```
+
+
+### Defining a dependency
+
+Defining a dependency will happen via a new hug `provides` decorator:
+
+```python
+    @hug.provides()
+    def mysql_connection(request, config):
+        ...
+```
+
+By default, the name will be the name of the function, but the first argument sent to that directive should enable redefining it:
+
+```python
+    @hug.provides("mysql")
+    def mysql_connection(request, config):
+        ...
+```
+
+It should also be possible to specify different dependency providers for different interfaces:
+
+```python
+    # Multiple interfaces using one provider
+    @hug.provides("mysql", interfaces=['http', 'websocket'])
+    def mysql_web():
+        ...
+
+
+    # For a provider for a single interface, this shorthand can be used
+    @hug.cli.provides("mysql")
+    def mysql_cli():
+        ...
+
+
+    # The following would also be a valid way to apply to multiple interfaces
+    @hug.http.provides("mysql")
+    @hug.cli.provides("mysql")
+    def mysql():
+        ...
+```
+
+### Using a dependency
+
+Dependencies will be pulled into individual endpoints via a `requires(` function, that will be called within function/method defaults:
+
+```python
+from hug import requires
+
+@hug.http()
+def hello_world(mysql=requires('mysql')):
+    pass
+```
+
+Arguments can be passed to the `requires` function, these will be passed directly to the `provides` function.
+This will be done in the same manner as Python's `funtools.partial` and follow the same function signature.
+Any non-provided arguments will then require the dependency to pull them from the current application action,
+like for normal hug calls, or via dependency injection. A key point: Any dependency can have unlimited sub-dependencies.
+
+Optionaly, you can also directly require any other normal Python callable:
+
+```python
+import json
+
+from hug import requires
+
+
+def shared_configuration(config_file_location="config_file.json"):
+    with open(config_file_location) as config_file:
+         return json.loads(config_file.read())
+
+
+@hug.http()
+def hello_world(config=requires(shared_configuration)):
+    pass
+```
```

Comment:
This will require some awesome documentation of API parameters to keep confusion to a minimum. Currently when looking at code I can know function parameter names are interface arguments if there is a `@hug.provides` or `@hug.{cli,http,local}` in front of it. Now any functions parameter naming can magically become runtime critical with no warning. (Looking at `def shared_configuration` as a programmer I have no warning that renaming `config_file_location` to `feature_x_config_file` will break my API interface behavior.)

Reply:
This is a good point. On the plus side, in your scenario,  you would get an error immediately by Python itself letting you know that you were referencing a non existent variable in the below `requires` function. It's also a fairly common pattern in python itself to use undecorated functions as callbacks in this way (for example: sort functions). Might think on this a bit longer.

Reply:
I don't think you would get the warning immediately. If I call the code from cron once a day with
```
$ http get http://localhost:8000/hello_world?config_file_location=my_config.json
```
and rename `def shared_configuration` `config_file_location` to `feature_x_config_file` I won't see the breakage until the next time cron runs.

Reply:
Because these requirements are defined at instantiation time, you would see an error when starting up the server, or even just importing the module, rather than at call time: https://pyfiddle.io/fiddle/ce3c1ff2-808d-4800-ac6b-ca5e13e5fb0c/?i=true

Reply:
I'm not talking about changing the `requires(shared_configuration)` call, but changing the name of a parameter inside of
```python
def shared_configuration(config_file_location="config_file.json"):
```
to
```python
def shared_configuration(feature_x_config_file="config_file.json"):
```

Reply:
Oh sorry for my misunderstanding! Your concern makes perfect sense

Reply:
Intelligent editors can trace the changing of a parameter name to find and edit all callers that explicitly set that parameter by keyword. But this change flows out to external interfaces outside of the python code and isn't easy to trace without a thorough understanding of hug and the program at hand.